bigquery/client.go

```diff
@@ -0,0 +1,216 @@
+	}
+
+	var err error
+	if bq.conf.ServiceAccountCreds == "" {
+		if bq.client, err = bigquery.NewClient(context.Background(), bq.conf.ProjectId); err != nil {
+			return nil, nil, err
+		}
+	} else if bq.conf.ServiceAccountCreds == "-" {
+		if bq.client, err = bigquery.NewClient(context.Background(), bq.conf.ProjectId, option.WithoutAuthentication()); err != nil {
+			return nil, nil, err
+		}
+	} else if !strings.HasPrefix(bq.conf.ServiceAccountCreds, "{") {
+		if bq.client, err = bigquery.NewClient(context.Background(), bq.conf.ProjectId, option.WithCredentialsFile(bq.conf.ServiceAccountCreds)); err != nil {
+			return nil, nil, err
+		}
+	} else {
+		if bq.client, err = bigquery.NewClient(context.Background(), bq.conf.ProjectId, option.WithCredentialsJSON([]byte(bq.conf.ServiceAccountCreds))); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	bq.dataset = bq.client.Dataset(bq.conf.DatasetName)
+	bq.table = bq.dataset.Table(bq.conf.TableName)
+
+	bq.uspClient, err = uspclient.NewClient(conf.ClientOptions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Parse the query interval
+	queryInterval, err := time.ParseDuration(conf.QueryInterval)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid query interval: %w", err)
+	}
+
+	chStopped := make(chan struct{})
+	bq.wg.Add(1)
+	go func() {
+		defer bq.wg.Done()
+		defer close(chStopped)
+
+		for {
+			err = bq.lookupAndSend(bq.ctx)
+			if err != nil || bq.conf.IsOneTimeLoad {
+				return
+			}
+
+			select {
+			case <-time.After(queryInterval):
+			case <-bq.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return bq, chStopped, nil
+}
+
+func (bq *BigQueryAdapter) lookupAndSend(ctx context.Context) error {
+	q := bq.client.Query(bq.conf.SqlQuery)
+	it, err := q.Read(ctx)
+	if err != nil {
+		return err
+	}
+
+	// used for accessing column names
+	tableRef := bq.client.DatasetInProject(bq.conf.BigQueryProject, bq.conf.DatasetName).Table(bq.conf.TableName)
+	meta, err := tableRef.Metadata(ctx)
+	if err != nil {
+		return err
+	}
+	schema := meta.Schema
+
+	errChan, rowsChan := make(chan error), make(chan []bigquery.Value, 5000)
+	go func() {
+		defer close(rowsChan)
+		for {
+			var row []bigquery.Value
+
+			err = it.Next(&row)
+			if errors.Is(err, iterator.Done) {
+				break
+			}
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			select {
+			case rowsChan <- row:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	doneChan := make(chan struct{})
+	go func() {
+		defer close(doneChan)
+		for row := range rowsChan {
+			// Check the context's Done channel before calling Next
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// Convert response to json format
+			rowMap := make(map[string]interface{})
+			for i, col := range row {
+				rowMap[schema[i].Name] = col // use column name to form json object
+			}
+
+			msg := &protocol.DataMessage{
+				JsonPayload: rowMap,
+				TimestampMs: uint64(time.Now().UnixNano() / int64(time.Millisecond)),
```

Comment:
Tiny thing but there is now a `time.Now().UnixMilli()` that does this in one shot.
```suggestion
				TimestampMs: uint64(time.Now().UnixMilli()),
```

Reply:
Nice, updated as suggested.